src/chains.rs

```diff
@@ -439,8 +435,39 @@ impl Rewrite for Chain {
         formatter.format_children(context, child_shape)?;
         formatter.format_last_child(context, shape, child_shape)?;
 
-        let result = formatter.join_rewrites(context, child_shape)?;
-        wrap_str(result, context.config.max_width(), shape)
+        formatter.join_rewrites(context, child_shape)
+    }
+}
+
+impl Rewrite for Chain {
+    fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String> {
+        debug!("rewrite chain {:?} {:?}", self, shape);
+
+        let result = self.format(context, shape)?;
+        if context.config.allow_chain_call_overflow() {
+            match context.config.indent_style() {
```

Comment:
I started out just returning `Some(result)` regardless of the indent style, but discovered that had what IMO was an undesirable side effect with `Visual` indent style when the lhs is long as well.  

It was resulting in:
```rust
fn foo() {
    let suuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuum = xxxxxxx
        .map(|x| x + 5)
        .map(|x| x / 2)
        .fold(0, |acc, x| acc + x);
}
```

being converted to 
```rust
fn foo() {
    let suuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuum = xxxxxxx.map(|x| x + 5)
                                                                          .map(|x| x / 2)
                                                                          .fold(0, |acc, x| acc + x);
}
```

instead of what `rustfmt` currently does
```rust
fn foo() {
    let suuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuum =
        xxxxxxx.map(|x| x + 5)
               .map(|x| x / 2)
               .fold(0, |acc, x| acc + x);
}
```

That happens because the line length was no longer being checked in `wrap_str`, and the start of the chain no longer being block indented. That would be _really_ bad if the call chain also included one or more long elements.

To address that, I've updated the flow for the `Visual` indent style to check the line length, and do a block indent if we're long. AFAICT this is working, and all the tests are passing, though I did also have to make one minor tweak in `exprs.rs` to prevent the `error[internal]: left behind trailing whitespace` error from occurring